app/src/main/java/android/coding/interview/makeitawesome/fragment/PicturesFragment.java

```diff
@@ -25,13 +46,104 @@ public static Fragment newInstance() {
     @Nullable
     @Override
     public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
-        RecyclerView rv = (RecyclerView) inflater.inflate(R.layout.pictures_list_fragment, container, false);
-        setupRecyclerView(rv);
-        return rv;
+        mPhotosAdapter = new PhotosAdapter(mData, getActivity());
+
+        httpGetTask = new HttpGetTask().execute(); //start background task
+
+        mRecyclerView = (RecyclerView) inflater.inflate(R.layout.pictures_list_fragment, container, false);
+        setupRecyclerView(mRecyclerView);
+        return mRecyclerView;
+    }
+
+    @Override
+    public void onPause() {
+        super.onPause();
+        if (dialog != null){
+            dialog.dismiss();
+        }
+        httpGetTask.cancel(true);
     }
 
     private void setupRecyclerView(RecyclerView recyclerView) {
+        mRecyclerView.addItemDecoration(new DividerItemDecoration(recyclerView.getContext(), DividerItemDecoration.VERTICAL_LIST));
+        mRecyclerView.setAdapter(mPhotosAdapter);
+        mPhotosAdapter.setClickListener(this);
         recyclerView.setLayoutManager(new LinearLayoutManager(recyclerView.getContext()));
-        recyclerView.setAdapter(new PhotosAdapter());
+    }
+
+    @Override
+    public void itemClicked(View view, int position) {
+        Intent intent = new Intent(getActivity(), Picture.class);
+        String url = mData.get(position).getUrl();
+
+        intent.putExtra(Picture.IMAGEURL, url);
+        startActivity(intent);
+    }
+
+    /**
+     * download the data
+     */
+    private class HttpGetTask extends AsyncTask<Void, Void, ArrayList<ImageData>> {
+        AndroidHttpClient mClient = AndroidHttpClient.newInstance("");
+
+
+        @Override
+        protected void onPreExecute() {
+            super.onPreExecute();
+            dialog = new ProgressDialog(getActivity());
+            dialog.setTitle(getString(R.string.loading));
+            dialog.setCancelable(false);
+            dialog.show();
+        }
+
+        @Override
+        protected ArrayList<ImageData> doInBackground(Void... unused) {
+            Log.v(TAG, "doInBackground");
+
+            HttpGet request = new HttpGet(URL);
+            JSONResponseHandler searchresponseHandler = new JSONResponseHandler();
+
+            try {
+                return mClient.execute(request, searchresponseHandler);
+            } catch (ClientProtocolException exception) {
+                exception.printStackTrace();
+            } catch (IOException ioexception) {
+                ioexception.printStackTrace();
+            } finally {
+                if (mClient != null) mClient.close();
+            }
+            return null;
+        }
+
+        @Override
+        protected void onPostExecute(ArrayList<ImageData> result) {
+            Log.v(TAG, "onPostExecute");
+            if (!isCancelled()){
+                dialog.dismiss();
```

Comment:
Dialog should be dimissed always.

No much to add to the improved version. He only included the progressdialog, basically.